### PR TITLE
Refector packages

### PR DESCRIPTION
**Pull request by Aflynn50**
This moves the contents of the `parse` and `typeinfo` packages to a new package called `expr`. All the functions changed are to be not exported so that they may be exported when needed.

The odd variable name has been changed here and there to avoid clashes with the new lowercase type.

### CODE REVIEW

File: internal/expr/parts.go [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+func (p *outputPart) toString() string {
+	return fmt.Sprintf("OutputPart[%+v %+v]", p.source, p.target)
+}
+
+func (p *outputPart) toSQL() string {
+	return ""
+}
+
+// BypassPart represents a part of the expression that we want to pass to the
```
Comment (TheSignPainter98) [inline internal/expr/parts.go]: `s/BypassPart/bypassPart/`

File: internal/expr/parts.go [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+// backend database verbatim.
+type bypassPart struct {
+	chunk string
+}
+
+func (p *bypassPart) toString() string {
+	return "BypassPart[" + p.chunk + "]"
```
Comment (TheSignPainter98) [inline internal/expr/parts.go]: Also here, `s/BypassPart/bypassPart/`

File: internal/expr/parts.go [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+}
+
+// InputPart represents a named parameter that will be sent to the database
+// while performing the query.
+type inputPart struct {
+	source fullName
+}
+
+func (p *inputPart) toString() string {
+	return fmt.Sprintf("InputPart[%+v]", p.source)
```
Comment (TheSignPainter98) [inline internal/expr/parts.go]: Capitalisation---should this be `inputPart[...]`?

File: internal/expr/parts.go [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+// OutputPart represents a named target output variable in the SQL expression,
+// as well as the source table and column where it will be read from.
+type outputPart struct {
+	source []fullName
+	target []fullName
+}
+
+func (p *outputPart) toString() string {
+	return fmt.Sprintf("OutputPart[%+v %+v]", p.source, p.target)
```
Comment (TheSignPainter98) [inline internal/expr/parts.go]: Likewise, should this be `outputPart`?

Comment (Aflynn50) [reply internal/expr/parts.go]: Changed.

File: internal/expr/parts.go (hunk 3 above)
Comment (Aflynn50) [reply internal/expr/parts.go]: Changed.

File: internal/expr/parts.go (hunk 2 above)
Comment (Aflynn50) [reply internal/expr/parts.go]: Changed.

File: internal/expr/parts.go (hunk 1 above)
Comment (Aflynn50) [reply internal/expr/parts.go]: Changed.

File: internal/expr/parser_test.go [hunk 5]
```diff
@@ -1,9 +1,8 @@
-package parse_test
+package expr
```
Comment (niemeyer) [inline internal/expr/parser_test.go]: PR looks good and seems ready to go in. As we discussed, this change is the only one that seems like a step backwards, as the tests were already using the public API.  If we revert this, we can merge the PR pretty much as-is and convert the other tests to also use the public API and poke holes through when actually required via the export_test.go pattern.

**Review by niemeyer — CHANGES_REQUESTED** [review]

**Review by niemeyer — APPROVED** [review]
Thanks for the changes.

**Pull request merged**